File: examples/readme/data-driven-generation/data_driven.py
"""Executable Python source for docs/features/data-driven-generation.md."""

# --8<-- [start:example]
from PIL import Image
from nucleation import Schematic


def barrel_position(x, y, channel):
    alternate = x & 1
    z = 5 + alternate if (channel & 1) == (y & 1) else 5 * alternate
    return (
        -((channel + y) & 1),
        -2 - channel - 3 * y,
        6 * (x // 2) + z - 2,
    )
rom = Schematic.create("image_rom")

with Image.open("rom-input.png") as source:
    image = source.convert("RGB")

    for index, (red, green, blue) in enumerate(image.getdata()):
        x, y = index % image.width, index // image.width

        for channel, signal in enumerate((blue >> 4, green >> 4, red >> 4)):
            rom.set_block(
                *barrel_position(x, y, channel),
                f"minecraft:barrel{{signal={signal}}}",
            )

rom.save_to_file("image-rom.schem")
# --8<-- [end:example]


assert rom.block_count() == 16 * 10 * 3
assert rom.tight_dimensions().x == 2
print("Data-driven Python example: OK")

File: examples/readme/data-driven-generation/data_driven.mjs
// --8<-- [start:example]
import { readFileSync, writeFileSync } from "node:fs";
import { PNG } from "pngjs";
import { Schematic } from "nucleation";


function barrelPosition(x, y, channel) {
  const alternate = x & 1;
  const z = (channel & 1) === (y & 1) ? 5 + alternate : 5 * alternate;
  return [
    -((channel + y) & 1),
    -2 - channel - 3 * y,
    6 * Math.floor(x / 2) + z - 2,
  ];
}
const image = PNG.sync.read(readFileSync("rom-input.png"));
const rom = Schematic.create("image_rom");

for (let y = 0; y < image.height; y += 1) {
  for (let x = 0; x < image.width; x += 1) {
    const offset = (y * image.width + x) * 4;
    const [red, green, blue] = image.data.subarray(offset, offset + 3);

    for (const [channel, signal] of [blue >> 4, green >> 4, red >> 4].entries()) {
      rom.setBlock(
        ...barrelPosition(x, y, channel),
        `minecraft:barrel{signal=${signal}}`,
      );
    }
  }
}

const bytes = Buffer.from(rom.toSchematicB64(), "base64");
writeFileSync("image-rom.schem", bytes);
// --8<-- [end:example]


if (rom.blockCount() !== 16 * 10 * 3) throw new Error("block count changed");
console.log("Data-driven JavaScript example: OK");
